Add SocketException to Abstractions-project

danielcweber opened the pull request:
In PCL-code, there is obviously no access to the platform specific SocketException so it may not be caught. Catching all exceptions is evil, so this branch adds a class called SocketException to the Abstractions-project. It's sole purpose is to wrap instances of the platform specific exception. I did this wrapping for the .net-implementation only for the moment until I have some feedback on this PR.

danielcweber committed:
.NET-implementation: Catch platform-specific SocketExceptions and wrap them.

rdavisau commented:
Hi again,

You're absolutely right - currently you have to resort to doing Bad Things without a way to test for the native exception classes, so this PR is great. My only reservation is that the code becomes a little noisy with the `try`/`catch` constructs everywhere, when we aren't really handling exceptions, just  'wrapping' them. To that I propose where possible we use something like a `WrapNativeSocketExceptions` extension on tasks in order to keep things more readable. I came up with the below, which looks to do the trick. 

```csharp
internal static class NativeExceptionExtensions
{
    internal static readonly HashSet<Type> NativeSocketExceptions = new HashSet<Type> { typeof(SocketException) };

    internal static Task WrapNativeSocketExceptions(this Task task)
    {
        return task
                .ContinueWith(t =>
                {
                    var ex = t.Exception.InnerException;

                    if (NativeSocketExceptions.Contains(ex.GetType()))
                        throw new PclSocketException(ex);
                    else
                        throw ex;
                }, TaskContinuationOptions.OnlyOnFaulted);
    }
}
```

so something like this:

```csharp
protected Task SendAsync(byte[] data)
{
  try
  {
      return _backingUdpClient.SendAsync(data, data.Length);
  }
  catch (PlatformSocketException ex)
  {
      throw new PclSocketException(ex);
  }
}
```
becomes:
```csharp
protected Task SendAsync(byte[] data)
{
    return _backingUdpClient.SendAsync(data, data.Length)
                            .WrapNativeSocketExceptions();
}
```
and then explicit try/catch calls are left for situations where we actually implement our own exception logic for some reason. I know there are a few exceptions <sub><sub>_ha ha_</sub></sub>  to that rule - situations where explicit `try`/`catch` is unavoidable like in the udp ctors - but broadly I think it is a good approach. 

Do you see any issues with that?

danielcweber commented:
Having this extension method is generally a good thing. However, as you said there are constructors and properties that have no async signature and still would need to be wrapped. You would end up with a mix of explicit try/catch blocks and usages of the extension. I don't know if that's what you yould want.

rdavisau commented:
Yeah, it's a good point, but I think I'm OK with it. I just browsed the WinRT implementation and on a quick assessment it looks like the non-Task exceptions do not apply on that side. Given that, I'm happy to include the udp ctor and tcpclient endpoint getter parts with explicit try/catch, and use the extension for the rest. Do you want to make the change? Otherwise I can merge and do from my end. There is a bit of time, I'm aiming to pack 1.3 with this and a couple of other bits around 20/June. :watch:

danielcweber commented:
Sorry for delay. I can make the change for .net as soon as I find some free time. I have no insights into WinRT however but I may give it a shot.

rdavisau commented:
No problems, I don't think I'll be ready by the 20th after all so there is still time. Thanks!

danielcweber committed:
Add NativeExceptionExtensions class that contains a method to to automatically wrap platform socket exceptions thrown by async methods.

danielcweber committed:
Use WrapNativeSocketExceptions where applicable in .NET-Implementation.

danielcweber commented:
Made the change for the .net implementation at last. I also rebased my branch on your current master. I was thinking about changing the extension method to something like this

    internal static class NativeExceptionExtensions
    {
        internal static readonly HashSet<Type> NativeSocketExceptions = new HashSet<Type> { typeof(SocketException) };

        internal static Exception WrapNativeSocketExceptions(this Exception ex)
        {
            if (NativeSocketExceptions.Contains(ex.GetType()))
                throw new PclSocketException(ex);
                
            throw ex;
        }
    }

Used like this:

    try
    {
        //do something
    }
    catch(Exception ex)
    {
        throw ex.WrapNativeSocketExceptions();
    }

That way, we would have try/catch blocks everywhere (even catching all exceptions at first) but not a mix.

rdavisau commented:
Thanks! I will do at the same for WinRT. When I see how it looks on WinRT, let's revisit the extension bit. The new proposal is an improvement from a readability/noise perspective. I will focus on sockets in my free for the next couple of days and finally get the other outstanding bits along with this ready for a packaging. Thanks again for the help.

rdavisau commented:
Ok so I finally got time to look over this properly and do the WinRT side. I `cherry-pick`ed your `SocketException` commit to reference it in my code but I think that has create a merge conflict, I thought it would Just Work. :( I will confirm whether that is the conflict a bit later. 

I noticed that in your branch, the `NativeExceptionExtensions` is in the Abstractions project. It should be in the .NET project, and the `SocketException` in the `HashSet` should be the `System.Net.Sockets` one. In the .NET project, you have access to the socket exception type to check whether it should be wrapped. I have done something [similar](https://github.com/rdavisau/sockets-for-pcl/blob/2a636561c45759bd8303c45c364e3d1b2ae120c1/Sockets/Sockets.Implementation.WinRT/Exceptions/NativeExceptionExtensions.cs) for WinRT based on what you had, but with the weirdness of WinRT. If you can move yours to the .NET project and check this [this](https://github.com/danielcweber/sockets-for-pcl/blob/e34bffb6b7d21f026e4139909ba61d049ae53313/Sockets/Sockets.Plugin.Abstractions/Extensions/NativeExceptionExtensions.cs#L42) which I think should be `OnlyOnFaulted`, then we are good to merge (besides the conflict.. of course). If you wanted to use the last exception helper you wrote for use in the non-task bits of code that can throw feel free. On WinRT it turned out that all native exceptions were on `Task`-based code. Thanks as always.

danielcweber commented:
Back from holidays, I will get back to this as soon as possible. Thanks for your patience.

rdavisau commented:
No problems; I ended up reverting my changes to pack some other fixes (and because I messed it up somehow) which means we can merge cleanly again too. I will sort WinRT out.

danielcweber commented:
I moved the extensions to the .net-platform project. And fixed the continuation option. Was there someting else?

If you add roughly the same code to the WinRT project, shoud we aim to reduce copy-paste and aim for more code reuse? I am currently looking for a clean way to reuse the extension and get the platform specific types into the HashSet.

danielcweber committed:
Remove a redundant if-statement.

rdavisau commented:
Hey @danielcweber, thanks for this. I'll probably not be able to look it over properly until after this weekend.

I might be thinking of the wrong thing, but for reuse we might be able to do something via the (by now, poorly-named) `Abstractions` project, which is shared by all platforms. I guess it it might require an initialisation step to set up, or we might be able to get away with a mini-bait and switch where the hashset lives in the platform projects but the extension lives in the shared code.

rdavisau commented:
Hi @danielcweber, 

I'm back and reviewing the WinRT side of this. I need to be a pain and ask you to resubmit this PR against the `dev` branch that I have just created. There are several changes I need to stage for the next release so I can't merge directly into `master` as I want it to continue to match the current release. I haven't been using a seperate branch in the past so it's my bad. Thanks again for your help.

danielcweber commented:
Rebased it onto dev. There's no way of changing the target branch of this PR so I opened a new one (#39)

rdavisau commented:
Great stuff. Thank you :+1:
